flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/FailureListener.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.executiongraph;
+
+import org.apache.flink.core.plugin.Plugin;
+import org.apache.flink.runtime.metrics.groups.JobManagerJobMetricGroup;
+
+/** Failure listener to customize the behavior for each type of failures tracked in job manager. */
+public interface FailureListener extends Plugin {
+
+    /**
+     * Initialize the listener with JobManagerJobMetricGroup.
+     *
+     * @param metricGroup metrics group that the listener can add customized metrics definition.
+     */
+    void init(JobManagerJobMetricGroup metricGroup);
+
+    /**
+     * Method to handle a failure in the listener.
+     *
+     * @param cause the failure cause
+     * @param globalFailure whether the failure is a global failure
+     */
+    void onFailure(final Throwable cause, boolean globalFailure);
```

Comment:
Shouldn't this also receive some information about the job? Given that the listener is configured per cluster, how would an implementation distinguish errors between different jobs within a session cluster?

I think it would be good to out-line the use-case that you're trying to accomplish; as it stands, outside of counting exceptions there's little you can do with the current interface.

Reply:
The `JobManagerJobMetricGroup` contains the JobId. Would it make sense to introduce a FailureListenerMetricGroup extends MetricGroup that exposes the JobId and JobName, and is not considered an internal API?

Reply:
The purpose of MetricGroups is not to make metadata readily accessible to other components.
Why not just pass the jobID/jobName/whatever metadata you linked separately into the constructor/init?

Reply:
Good point, thanks for the explanation.

Reply:
+1